README.md
# Welcome to libarchive!

The libarchive project develops a portable, efficient C library that
can read and write streaming archives in a variety of formats.  It
also includes implementations of the common `tar`, `cpio`, and `zcat`
command-line tools that use the libarchive library.

## Questions?  Issues?

* http://www.libarchive.org is the home for ongoing
  libarchive development, including documentation,
  and links to the libarchive mailing lists.
* To report an issue, use the issue tracker at
  https://github.com/libarchive/libarchive/issues
* To submit an enhancement to libarchive, please
  submit a pull request via GitHub: https://github.com/libarchive/libarchive/pulls

## Contents of the Distribution

This distribution bundle includes the following major components:

* **libarchive**: a library for reading and writing streaming archives
* **tar**: the 'bsdtar' program is a full-featured 'tar' implementation built on libarchive
* **cpio**: the 'bsdcpio' program is a different interface to essentially the same functionality
* **cat**: the 'bsdcat' program is a simple replacement tool for zcat, bzcat, xzcat, and such
* **examples**: Some small example programs that you may find useful.
* **examples/minitar**: a compact sample demonstrating use of libarchive.
* **contrib**:  Various items sent to me by third parties; please contact the authors with any questions.

The top-level directory contains the following information files:

* **NEWS** - highlights of recent changes
* **COPYING** - what you can do with this
* **INSTALL** - installation instructions
* **README** - this file
* **CMakeLists.txt** - input for "cmake" build tool, see INSTALL
* **configure** - configuration script, see INSTALL for details.  If your copy of the source lacks a `configure` script, you can try to construct it by running the script in `build/autogen.sh` (or use `cmake`).

The following files in the top-level directory are used by the 'configure' script:
* `Makefile.am`, `aclocal.m4`, `configure.ac` - used to build this distribution, only needed by maintainers
* `Makefile.in`, `config.h.in` - templates used by configure script

## Documentation

In addition to the informational articles and documentation
in the online [libarchive Wiki](https://github.com/libarchive/libarchive/wiki),
the distribution also includes a number of manual pages:

 * bsdtar.1 explains the use of the bsdtar program
 * bsdcpio.1 explains the use of the bsdcpio program
 * bsdcat.1 explains the use of the bsdcat program
 * libarchive.3 gives an overview of the library as a whole
 * archive_read.3, archive_write.3, archive_write_disk.3, and
   archive_read_disk.3 provide detailed calling sequences for the read
   and write APIs
 * archive_entry.3 details the "struct archive_entry" utility class
 * archive_internals.3 provides some insight into libarchive's
   internal structure and operation.
 * libarchive-formats.5 documents the file formats supported by the library
 * cpio.5, mtree.5, and tar.5 provide detailed information about these
   popular archive formats, including hard-to-find details about
   modern cpio and tar variants.

The manual pages above are provided in the 'doc' directory in
a number of different formats.

You should also read the copious comments in `archive.h` and the
source code for the sample programs for more details.  Please let us
know about any errors or omissions you find.

## Supported Formats

Currently, the library automatically detects and reads the following fomats:
  * Old V7 tar archives
  * POSIX ustar
  * GNU tar format (including GNU long filenames, long link names, and sparse files)
  * Solaris 9 extended tar format (including ACLs)
  * POSIX pax interchange format
  * POSIX octet-oriented cpio
  * SVR4 ASCII cpio
  * Binary cpio (big-endian or little-endian)
  * ISO9660 CD-ROM images (with optional Rockridge or Joliet extensions)
  * ZIP archives (with uncompressed or "deflate" compressed entries, including support for encrypted Zip archives)
  * GNU and BSD 'ar' archives
  * 'mtree' format
  * 7-Zip archives
  * Microsoft CAB format
  * LHA and LZH archives
  * RAR and RAR 5.0 archives (with some limitations due to RAR's proprietary status)
  * XAR archives

The library also detects and handles any of the following before evaluating the archive:
  * uuencoded files
  * files with RPM wrapper
  * gzip compression
  * bzip2 compression
  * compress/LZW compression
  * lzma, lzip, and xz compression
  * lz4 compression
  * lzop compression
  * zstandard compression

The library can create archives in any of the following formats:
  * POSIX ustar
  * POSIX pax interchange format
  * "restricted" pax format, which will create ustar archives except for
    entries that require pax extensions (for long filenames, ACLs, etc).
  * Old GNU tar format
  * Old V7 tar format
  * POSIX octet-oriented cpio
  * SVR4 "newc" cpio
  * shar archives
  * ZIP archives (with uncompressed or "deflate" compressed entries)
  * GNU and BSD 'ar' archives
  * 'mtree' format
  * ISO9660 format
  * 7-Zip archives
  * XAR archives

When creating archives, the result can be filtered with any of the following:
  * uuencode
  * gzip compression
  * bzip2 compression
  * compress/LZW compression
  * lzma, lzip, and xz compression
  * lz4 compression
  * lzop compression
  * zstandard compression

## Notes about the Library Design

The following notes address many of the most common
questions we are asked about libarchive:

* This is a heavily stream-oriented system.  That means that
  it is optimized to read or write the archive in a single
  pass from beginning to end.  For example, this allows
  libarchive to process archives too large to store on disk
  by processing them on-the-fly as they are read from or
  written to a network or tape drive.  This also makes
  libarchive useful for tools that need to produce
  archives on-the-fly (such as webservers that provide
  archived contents of a users account).

* In-place modification and random access to the contents
  of an archive are not directly supported.  For some formats,
  this is not an issue: For example, tar.gz archives are not
  designed for random access.  In some other cases, libarchive
  can re-open an archive and scan it from the beginning quickly
  enough to provide the needed abilities even without true
  random access.  Of course, some applications do require true
  random access; those applications should consider alternatives
  to libarchive.

* The library is designed to be extended with new compression and
  archive formats.  The only requirement is that the format be
  readable or writable as a stream and that each archive entry be
  independent.  There are articles on the libarchive Wiki explaining
  how to extend libarchive.

* On read, compression and format are always detected automatically.

* The same API is used for all formats; it should be very
  easy for software using libarchive to transparently handle
  any of libarchive's archiving formats.

* Libarchive's automatic support for decompression can be used
  without archiving by explicitly selecting the "raw" and "empty"
  formats.

* I've attempted to minimize static link pollution.  If you don't
  explicitly invoke a particular feature (such as support for a
  particular compression or format), it won't get pulled in to
  statically-linked programs.  In particular, if you don't explicitly
  enable a particular compression or decompression support, you won't
  need to link against the corresponding compression or decompression
  libraries.  This also reduces the size of statically-linked
  binaries in environments where that matters.

* The library is generally _thread safe_ depending on the platform:
  it does not define any global variables of its own.  However, some
  platforms do not provide fully thread-safe versions of key C library
  functions.  On those platforms, libarchive will use the non-thread-safe
  functions.  Patches to improve this are of great interest to us.

* In particular, libarchive's modules to read or write a directory
  tree do use `chdir()` to optimize the directory traversals.  This
  can cause problems for programs that expect to do disk access from
  multiple threads.  Of course, those modules are completely
  optional and you can use the rest of libarchive without them.

* The library is _not_ thread aware, however.  It does no locking
  or thread management of any kind.  If you create a libarchive
  object and need to access it from multiple threads, you will
  need to provide your own locking.

* On read, the library accepts whatever blocks you hand it.
  Your read callback is free to pass the library a byte at a time
  or mmap the entire archive and give it to the library at once.
  On write, the library always produces correctly-blocked output.

* The object-style approach allows you to have multiple archive streams
  open at once.  bsdtar uses this in its "@archive" extension.

* The archive itself is read/written using callback functions.
  You can read an archive directly from an in-memory buffer or
  write it to a socket, if you wish.  There are some utility
  functions to provide easy-to-use "open file," etc, capabilities.

* The read/write APIs are designed to allow individual entries
  to be read or written to any data source:  You can create
  a block of data in memory and add it to a tar archive without
  first writing a temporary file.  You can also read an entry from
  an archive and write the data directly to a socket.  If you want
  to read/write entries to disk, there are convenience functions to
  make this especially easy.

* Note: The "pax interchange format" is a POSIX standard extended tar
  format that should be used when the older _ustar_ format is not
  appropriate.  It has many advantages over other tar formats
  (including the legacy GNU tar format) and is widely supported by
  current tar implementations.

test


NEWS
Jul 09, 2017: libarchive 3.3.2 released

dummy change 1 2 3 4 5

Mar 16, 2017: NFSv4 ACL support for Linux (librichacl)

Feb 26, 2017: libarchive 3.3.1 released
    Security & Feature release

Feb 19, 2017: libarchive 3.3.0 released
    Security & Feature release

Jan 29, 2017: Limited NFSv4 ACL support for Mac OS (Darwin)

Jan 10, 2017: POSIX.1e and NFSv4 ACL support for Solaris and derivates

Dec 27, 2016: NFSv4 ACL read and write support for pax
    Deprecated functions: archive_entry_acl_text(), archive_entry_acl_text_w()

Nov, 2016: libarchive is now being tested by the OSS-Fuzz project

Oct 26, 2016: Remove liblzmadec support

Oct 23, 2016: libarchive 3.2.2 released
    Security release

Jun 20, 2016: libarchive 3.2.1 released
    This fixes a handful of security and other critical issues with 3.2.0

May 01, 2016: libarchive 3.2.0 released

Apr 09, 2016: libarchive 3.1.901a released
    Another test release in preparation for 3.2.0

Feb 13, 2016: libarchive 3.1.900a released
    This is a test release in preparation for 3.2.0

Oct 21, 2015: Preliminary port to OSF

Apr 11, 2015: libarchive's issue tracker is now hosted at GitHub.
    https://github.com/libarchive/libarchive/issues

Early 2015:  Many fixes to crash and overflow bugs thanks to Hanno Boeck

Oct 13, 2014: Zip encryption and decryption support

Aug 13, 2014: Add support for lz4 compression.

Jun 10, 2014: Add warc format support

May  3, 2014: Add experimental Zip streaming extension

Apr  6, 2014: Add bsdcat command-line tool

Jan 12, 2014: Add Zip64 support

Dec  1, 2013: Rewrite Zip write logic

Jul  1, 2013: Add ability to detect encrypted entries for many formats
     (This does not add the ability to *decrypt* those entries, however)

Feb 23, 2013: "raw" write support added

Feb 09, 2013: libarchive 3.1.2 released

Jan 28, 2013: libarchive's new website moved to http://www.libarchive.org.

Jan 13, 2013: libarchive 3.1.1 released

Jan 13, 2013: libarchive 3.1.0 released

Dec 07, 2012: Implement functions to manually set the format and filters used.

Nov 11, 2012: Add support for __MACOSX directory in Zip archives, which resource
    forks are stored in.

Oct 20, 2012: Add support for writing v7 tar format.

Oct 09, 2012: Add support for grzip compression.

Oct 07, 2012: Introduce b64encode filter.
Oct 07, 2012: Introduce uuencode filter.

Oct 06, 2012: Add support for lzop.

Sep 27, 2012: Implement function used to seek within data blocks.
    (Currently only supported for uncompressed RAR archives).

Apr 22, 2012: Add basic archive read and write filter support for lrzip.

Mar 27, 2012: libarchive 3.0.4 released

Feb 05, 2012: libarchive development now hosted at GitHub.
    http://libarchive.github.com/
Feb 05, 2012: libarchive's issue tracker remains at Google Code.
    http://code.google.com/p/libarchive/issues/list
Feb 05, 2012: libarchive's mailing lists remain at Google Groups.

Dec 24, 2011: libarchive 3.0.2 released
Dec 23, 2011: Various fixes merged from FreeBSD
Dec 23, 2011: Symlink support in Zip reader and writer
Dec 23, 2011: Robustness fixes to 7Zip reader

Nov 27, 2011: libarchive 3.0.1b released

Nov 26, 2011: 7Zip reader
Nov 26, 2011: Small fixes to ISO and Zip to improve robustness with corrupted input
Nov 24, 2011: Improve streaming Zip reader's support for uncompressed entries
Nov 20, 2011: New seeking Zip reader supports SFX Zip archives
Nov 20, 2011: Build fixes on Windows

Nov 13, 2011: libarchive 3.0.0a released

Nov 06, 2011: Update shared-library version calculations for libarchive 3.x
Sep 04, 2011: Fix tar -s; follow GNU tar for controlling hardlink/symlink substitutions
Aug 18, 2011: Fix reading ISO images built by NetBSD's mkisofs
Aug 15, 2011: Old archive_read_support_compression_XXX functions are deprecated and
   will disappear in libarchive 4.0.
Jun 26, 2011: RAR reader
Jun 16, 2011: Add tar:compat-2x option to emulate broken libarchive 2.x
   handling of pax UTF-8 headers
Apr 25, 2011: Refactor read_open() into a collection of single-item setters;
   support the old interfaces as wrappers
Apr 12, 2011: Split disk writer into separate POSIX and Windows implementations
Apr 10, 2011: Improvements to character translations on Windows.
Mar 30, 2011: More work to return errors instead of calling abort()
Mar 23, 2011: Add charset option to many writers to control MBCS filenames
Mar 17, 2011: Overhauled support for per-format extension options
Mar 17, 2011: Track character set used for mbcs strings, support
   translating to/from user-specified locale
Mar 09, 2011: Recognize mtree files without requiring a signature
Mar 06, 2011: Use iconv to convert to/from Unicode instead of making bad
   assumptions about the C90 character set translation functions
Feb 17, 2011: Fixes for AIX, TRU64, and other platforms
Dec 22, 2010: CAB reader
Dec 20, 2010: LHA/LZH reader
Jul 03, 2010: minitar example demonstrates archive_read_disk directory traversal
Jun 29, 2010: Many improvements to ISO reader compatibility
Jun 26, 2010: Use larger buffers when copy files into an archive
Jun 18, 2010: Reimplement Mac OS extensions in libarchive
Jun 09, 2010: archive_read_disk now supports traversals
May 28, 2010: XAR writer
May 16, 2010: Fix ^T handling; don't exit on interrupted reads and writes
May 09, 2010: Improved detection of platform-specific crypto support
May 04, 2010: lzip read and write filters
May 01, 2010: New options: tar --gid --gname --uid --uname
Apr 28, 2010: Use Red-black tree for ISO reader/writer to improve performance
Apr 17, 2010: Minimal writer for legacy GNU tar format
Mar 12, 2010: Don't dereference symlinks on Linux when reading ACLs.
Mar 06, 2010: Fix build when an older libarchive is already installed
Feb 28, 2010: Relax handling of state failures; misuse by clients now generally
      results in a sticky ARCHIVE_FATAL rather than a visit to abort()
Feb 25, 2010: ISO writer
Feb 21, 2010: Split many man pages into smaller chunks.

libarchive/archive_inflate64.h
/*-
 * Copyright (c) 2018 Grzegorz Antoniak (http://antoniak.org)
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef __ARCHIVE_INFLATE64_H_
#define __ARCHIVE_INFLATE64_H_

#include <stdint.h>
#include <stdlib.h>

struct inflate64internal;

enum INF64_RET {
    INF64_OK = 0, INF64_ALLOC = 1, INF64_ERROR = 2, INF64_ARG = 3, INF64_IO = 4,
    INF64_BADDATA = 5, INF64_NEEDMORE = 6,
};

struct inflate64stream {
    /* API */
    ssize_t avail_in;
    ssize_t avail_out;
    ssize_t total_in;
    ssize_t total_out;
    const uint8_t* next_in;
    uint8_t* next_out;

    /* Not API */
    struct inflate64internal* inner;
    uint32_t initialized;
};

int inflate64init(struct inflate64stream* s);
int inflate64run(struct inflate64stream* s);
int inflate64finish(struct inflate64stream* s);

#endif


libarchive/archive_inflate64.c
/*-
 * Copyright (c) 2018 Grzegorz Antoniak (http://antoniak.org)
 * All rights reserved.
 *
 * This zlib-free deflate64 decoder is based on 7-Zip's deflate64 
 * decoder by Igor Pavlov and Info-ZIP's inflate.c by Mark Adler.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "archive_inflate64.h"

#include <stdlib.h>
#include <string.h>

#include <stdio.h>
#define LOG(...) do { printf(__VA_ARGS__); puts(""); } while(0)
#define HEXDUMP(what, size) do {\
        size_t __i = 0;\
        uint8_t* __x = (uint8_t*) &what;\
        for(__i = 0; __i < (size_t) size; __i++) {\
            printf("%02x ", __x[__i]);\
        }\
        \
        printf("\n");\
    } while(0)
#define HEXDUMP_PTR(what, size) do {\
        size_t __i = 0;\
        const uint8_t* __x = (const uint8_t*) what;\
        for(__i = 0; __i < (size_t) size; __i++) {\
            printf("%02x ", __x[__i]);\
        }\
        \
        printf("\n");\
    } while(0)
#define INF64_TRY(x) do { int __ret; if(INF64_OK != (__ret = (x))) { LOG("TRY: error %d at %s:%d", __ret, __FILE__, __LINE__); return __ret; } } while(0)
#define INF64_TRY_OR(x, ref) do { int __ret; if(INF64_OK != (__ret = (x))) { return __ret; } else { ref = __ret; } } while(0)
#define INF64_COUNT_OF(x) (sizeof(x) / sizeof(*x))

const uint32_t INIT_TAG = 0xa1a2a3a4;

#define CODE_LENGTH_TABLE_SIZE 19
#define END_OF_BLOCK_SYMBOL    0x100
#define MATCH_SYMBOL           (END_OF_BLOCK_SYMBOL + 1)
#define LEN_TABLE_SIZE         31
#define DIST_TABLE_SIZE        32
#define MAIN_TABLE_SIZE        (LEN_TABLE_SIZE + MATCH_SYMBOL)

struct bit_reader {
    uint8_t pos;
    uint32_t cache;
    uint32_t value;
    uint8_t invert_table[256];
};

struct huff_table {
    uint32_t ll;
    uint32_t dl;
    uint32_t lc;
    uint8_t levels[CODE_LENGTH_TABLE_SIZE];
    uint8_t lens_7bit[1 << 7]; /* 128 */
    uint8_t ll_tab[MAIN_TABLE_SIZE];
    uint8_t dl_tab[DIST_TABLE_SIZE];
};

struct decoder_ctx {
    uint16_t lens_9bit[1 << 9]; /* 512 */
    uint16_t symtab[MAIN_TABLE_SIZE];
    uint32_t limtab[17];
    uint32_t postab[16];
};

const uint8_t code_len_pos[CODE_LENGTH_TABLE_SIZE] = {
    16, 17, 18, 0, 8, 7, 9, 6, 10, 5, 11, 4, 12, 3, 13, 2, 14, 1, 15
};

struct inflate64internal {
    struct bit_reader br;
    struct huff_table huff;
    struct decoder_ctx dec;
};

static
void br_reset(struct inflate64stream* s) {
    struct bit_reader* const br = &s->inner->br;
    int i;

    br->pos = 32;

    /* Bit reversal lookup table initialization. */
    for(i = 0; i < 256; i++) {
        uint32_t x = ((i & 0x55) << 1) | ((i & 0xAA) >> 1);
        x = ((x & 0x33) << 2) | ((x & 0xCC) >> 2);
        x = (uint8_t) (((x & 0x0F) << 4) | ((x & 0xF0) >> 4));
        br->invert_table[i] = x;
    }
}

/*static*/
/*int br_check_avail(struct inflate64stream* s, uint32_t how_many) {*/
    /*if(s->total_in >= s->avail_in)*/
        /*return INF64_NEEDMORE;*/

    /*if(s->avail_in - s->total_in < how_many)*/
        /*return INF64_NEEDMORE;*/

    /*return INF64_OK;*/
/*}*/

static
int br_fill_raw(struct inflate64stream* s, uint8_t* pbyte) {
    uint8_t byte = 0;

    if(s->total_in >= s->avail_in)
        return INF64_IO;

    byte = s->next_in[s->total_in++];
    *pbyte = byte;
    return INF64_OK;
}

static
int br_fill(struct inflate64stream* s) {
    struct bit_reader* const br = &s->inner->br;

    while(br->pos >= 8) {
        uint8_t b = 0;

        INF64_TRY(br_fill_raw(s, &b));

        br->cache = br->cache | (((uint32_t) b) << (32 - br->pos));
        br->value = (br->value << 8) | br->invert_table[b];
        br->pos -= 8;
    }

    return INF64_OK;
}

static inline
void br_skip(struct inflate64stream* s, int n) {
    struct bit_reader* const br = &s->inner->br;
    br->pos += n;
    br->cache >>= n;
}

// n is up to 32
// naming comes from [bitreader]_read_[n=multiple bytes]
static
int br_read_n(struct inflate64stream* s, int n, uint32_t* pvalue) {
    struct bit_reader* const br = &s->inner->br;

    INF64_TRY(br_fill(s));

    *pvalue = br->cache & ((1 << n) - 1);
    br_skip(s, n);
    return INF64_OK;
}

static
int br_readval_n(struct inflate64stream* s, int n, uint32_t* pvalue) {
    struct bit_reader* const br = &s->inner->br;
    const uint32_t mask = (1 << 24) - 1;

    INF64_TRY(br_fill(s));

    *pvalue = ((br->value >> (8 - br->pos)) & mask) >> (24 - n);
    return INF64_OK;
}

// n is up to 8
// convinience function that allows to read directly to uint8_t* without
// triggering any type errors
// naming comes from [bitreader]_read_[n=multiple bytes]_[uint8_t][pointer]
static inline
int br_read_n_u8p(struct inflate64stream* s, int n, uint8_t* pvalue) {
    uint32_t buf;
    INF64_TRY(br_read_n(s, n, &buf));
    *pvalue = (uint8_t) (buf & 255);
    return INF64_OK;
}

static
int valid_context(struct inflate64stream* s) {
    int ret = s->initialized == INIT_TAG && s->inner;
    return ret ? INF64_OK : INF64_ARG;
}

int inflate64init(struct inflate64stream* s) {
    /* Don't allow re-initialization of already initialized instance. */
    if(INF64_OK == valid_context(s)) {
        return INF64_ARG;
    }

    s->inner = (struct inflate64internal*) malloc(sizeof(struct inflate64internal));
    if(!s->inner)
        return INF64_ALLOC;

    memset(s->inner, 0, sizeof(struct inflate64internal));
    br_reset(s);
    s->initialized = INIT_TAG;
    return INF64_OK;
}

static
int levels_load(struct inflate64stream* s) {
    size_t i;
    struct huff_table* huff = &s->inner->huff;

    LOG("reading huffman dynamic table");

    INF64_TRY(br_read_n(s, 5, &huff->ll));
    INF64_TRY(br_read_n(s, 5, &huff->dl));
    INF64_TRY(br_read_n(s, 4, &huff->lc));

    huff->ll += 257;
    huff->dl += 1;
    huff->lc += 4;

    LOG("%d/%d/%d", huff->ll, huff->dl, huff->lc);

    for(i = 0; i < INF64_COUNT_OF(huff->levels); ++i) {
        const size_t position = code_len_pos[i];
        if(i < huff->lc) {
            INF64_TRY(br_read_n_u8p(s, 3, &huff->levels[position]));
        } else {
            huff->levels[position] = 0;
        }
    }

    for(i = 0; i < INF64_COUNT_OF(huff->levels); ++i) {
        LOG("1 levels[%zu] = %02x", i, huff->levels[i]);
    }

    return INF64_OK;
}

static
int levels_init(struct inflate64stream* s) {
    size_t i;
    ssize_t k;
    struct huff_table* const huff = &s->inner->huff;
    
    uint32_t len_counts[8] = { 0 };
    uint32_t temp_postab[8] = { 0 };
    uint32_t postab[8] = { 0 };
    uint32_t limits[8] = { 0 };
    uint32_t sym = 0;
    uint32_t start_pos = 0;

    for(sym = 0; sym < INF64_COUNT_OF(huff->levels); ++sym) {
        const size_t pos = huff->levels[sym];
        ++len_counts[pos];
    }

    len_counts[0] = 0;
    postab[0] = 0;
    limits[0] = 0;

    for(i = 1; i < 8; ++i) {
        start_pos += len_counts[i] << (7 - i);
        if(start_pos > (1 << 7))
            return INF64_BADDATA;

        limits[i] = start_pos;
        postab[i] = postab[i - 1] + len_counts[i - 1];
        temp_postab[i] = postab[i];
    }

    for(sym = 0; sym < INF64_COUNT_OF(huff->levels); ++sym) {
        const uint8_t len = huff->levels[sym];
        uint32_t offset;
        uint8_t val;
        size_t index;

        if(len == 0)
            continue;

        /* 'len' will never be more than 7, because it's read from the
         * file by using only 3 bits. */

        offset = temp_postab[len];
        ++temp_postab[len];
        offset -= postab[len];

        val = (uint8_t) ((sym << 3) | len);
        index = limits[len - 1] + (offset << (7 - len));
        for(k = 0; k < (1 << (7 - len)); ++k) {
            if((index + k) > (ssize_t) INF64_COUNT_OF(huff->lens_7bit)) {
                LOG("fatal: index+k = %zu, should be max %zu", index + k, INF64_COUNT_OF(huff->lens_7bit));
                return INF64_BADDATA;
            }

            LOG("save val %02x to index+k=%zu", val, index+k);
            huff->lens_7bit[index + k] = val;
        }
    }

    for(k = 0; k < (128 - limits[7]); ++k) {
        if((limits[7] + k) > (ssize_t) INF64_COUNT_OF(huff->lens_7bit)) {
            LOG("fatal: limits[7]+k = %zu, should be max %zu", limits[7] + k, INF64_COUNT_OF(huff->lens_7bit));
            return INF64_BADDATA;
        }

        huff->lens_7bit[limits[7] + k] = 0x1F << 3;
    }

    for(sym = 0; sym < 128; sym++) {
        LOG("lens_7bit[%d]=%d", sym, huff->lens_7bit[sym]);
    }

    return INF64_OK;
}

static
int levels_decode(struct inflate64stream* s, uint8_t* levels, int sym_count) {
    struct huff_table* const huff = &s->inner->huff;
    ssize_t i = 0;
    uint32_t num_bits = 0, num = 0, bit_data = 0;
    uint8_t symbol = 0;

    uint32_t val, pair, len, sym;

    do {
        INF64_TRY(br_readval_n(s, 7, &val));
        pair = huff->lens_7bit[val];

        len = pair & 7;
        br_skip(s, len);
        sym = pair >> 3;

        LOG("sym=0x%08x", sym);

        if(sym < 16) {
            levels[i++] = (uint8_t) (sym & 255);
            continue;
        }

        if(sym >= CODE_LENGTH_TABLE_SIZE) {
            return INF64_BADDATA;
        }

        if(sym == 16) {
            if(i == 0) {
                return INF64_BADDATA;
            }

            num_bits = 2;
            num = 0;
            symbol = levels[i - 1];
        } else {
            sym = (sym - 17) << 2;
            num_bits = 3 + sym;
            num = (uint32_t) sym << 1;
            symbol = 0;
        }

        INF64_TRY(br_read_n(s, num_bits, &bit_data));

        num = num + 3 + i + bit_data;
        if(num > (uint32_t) sym_count) {
            return INF64_BADDATA;
        } 

        do {
            levels[i++] = symbol;
        } while(i < num);
    } while(i < sym_count);

    return INF64_OK;
}

static
int decoder_init(struct inflate64stream* s) {
    struct huff_table* huff = &s->inner->huff;
    struct decoder_ctx* dec = &s->inner->dec;
    uint32_t len_counts[INF64_COUNT_OF(dec->postab)];
    uint32_t tmp_postab[INF64_COUNT_OF(dec->postab)];
    uint32_t i, sym, start_pos = 0;

    // kNumBitsMax = 15
    // m_numSymbols = MAIN_TABLE_SIZE
    // kNumTableBits = 9

    memset(len_counts, 0, sizeof(len_counts));

    for(sym = 0; sym < MAIN_TABLE_SIZE; ++sym) {
        const size_t pos = huff->ll_tab[sym];
        len_counts[pos]++;
    }

    len_counts[0] = 0;
    dec->postab[0] = 0;
    dec->limtab[0] = 0;

    for(i = 1; i < 15; ++i) {
        start_pos = start_pos + (len_counts[i] << (15 - i));
        if(start_pos > (1 << 15)) {
            LOG("Wrong start_pos");
            return INF64_BADDATA;
        }

        dec->limtab[i] = start_pos;
        dec->postab[i] = dec->postab[i - 1] + len_counts[i - 1];
        tmp_postab[i] = dec->postab[i];
    }

    dec->limtab[16] = 1 << 15;

    for(sym = 0; sym < MAIN_TABLE_SIZE; ++sym) {
        uint32_t offset;
        uint32_t len = huff->ll_tab[sym];
        if(len == 0)
            continue;

        offset = tmp_postab[len];
        dec->symtab[offset] = (uint16_t) sym;
        tmp_postab[len] = offset + 1;

        LOG("len=%d, offset=%d, sym=%d", len, offset, sym);

        if(len <= 9) {
            uint32_t num, k;
            uint16_t val;

            offset = offset - dec->postab[len];
            num = (uint32_t) 1 << (9 - len);
            val = (uint16_t) ((sym << 4) | len);
            LOG("num=%d, val=%d", num, val);
            k = (dec->limtab[len - 1] >> (15 - 9)) + (offset << (9 - len));

            for(i = 0; i < num; i++) {
                dec->lens_9bit[i + k] = val;
            }
        }
    }

    for(i = 0; i < (1 << 9); i++) {
        LOG("lens_9bit[%d]=%d", i, dec->lens_9bit[i]);
    }

    return INF64_OK;
}

static
int read_table_huff_dyn(struct inflate64stream* s) {
    struct huff_table* huff = &s->inner->huff;
    uint8_t temp_levels[MAIN_TABLE_SIZE + DIST_TABLE_SIZE];

    INF64_TRY(levels_load(s));
    INF64_TRY(levels_init(s));
    INF64_TRY(levels_decode(s, temp_levels, huff->ll + huff->dl));

    memset(huff->ll_tab, 0, MAIN_TABLE_SIZE);
    memcpy(huff->ll_tab, temp_levels, huff->ll);
    memset(huff->dl_tab, 0, DIST_TABLE_SIZE);
    memcpy(huff->dl_tab, &temp_levels[huff->ll], huff->dl);

    INF64_TRY(decoder_init(s));

    return INF64_OK;
}

int inflate64run(struct inflate64stream* s) {
    uint32_t final_block = 0;
    uint32_t block_type = 0;

    INF64_TRY(valid_context(s));
    INF64_TRY(br_read_n(s, 1, &final_block));
    INF64_TRY(br_read_n(s, 2, &block_type));

    enum {
        HUFFMAN_STORED = 0,
        HUFFMAN_STATIC = 1,
        HUFFMAN_DYNAMIC = 2,
    };

    LOG("bit: %d", final_block);

    switch(block_type) {
        case HUFFMAN_STORED:
            LOG("stored huffman tables are not supported yet");
            return INF64_BADDATA;
        case HUFFMAN_STATIC:
            LOG("static huffman tables are not supported yet");
            return INF64_BADDATA;
        case HUFFMAN_DYNAMIC:
            INF64_TRY(read_table_huff_dyn(s));
            break;
        default:
            LOG("got block_type=%d, which is not supported", block_type);
            return INF64_BADDATA;
    }

    LOG("ok");
    return INF64_OK;
}

int inflate64finish(struct inflate64stream* s) {
    INF64_TRY(valid_context(s));

    free(s->inner);
    s->inner = NULL;
    s->initialized = 0;

    return INF64_OK;
}
